src/main/java/org/traccar/protocol/TtnHttpProtocolDecoder.java

```diff
@@ -0,0 +1,258 @@
+import java.net.SocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TtnHttpProtocolDecoder extends BaseHttpProtocolDecoder {
+
+    public TtnHttpProtocolDecoder(Protocol protocol) {
+        super(protocol);
+    }
+
+    @Override
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        FullHttpRequest request = (FullHttpRequest) msg;
+        String contentType = request.headers().get(HttpHeaderNames.CONTENT_TYPE);
+        if (contentType != null && contentType.startsWith(HttpHeaderValues.APPLICATION_JSON.toString())) {
+            return decodeJson(channel, remoteAddress, request);
+        } else {
+            sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+            return null;
+        }
+    }
+
+    private void positionSetFromJson(Position position, String key, JsonValue value) {
+        switch (value.getValueType()) {
+            case STRING:
+                position.set(key, ((JsonString) value).getString());
+                break;
+            case NUMBER:
+                position.set(key, ((JsonNumber) value).doubleValue());
+                break;
+            case TRUE:
+                position.set(key, true);
+                break;
+            case FALSE:
+                position.set(key, false);
+                break;
+            default:
+                // Ignore other types
+                break;
+        }
+    }
+
+    private Object decodeJson(
+            Channel channel, SocketAddress remoteAddress, FullHttpRequest request) throws Exception {
+
+        String content = request.content().toString(StandardCharsets.UTF_8);
+        JsonObject root = Json.createReader(new StringReader(content)).readObject();
+
+        DeviceSession deviceSession = getDeviceSession(channel, remoteAddress,
+                root.getJsonObject("end_device_ids").getString("device_id"));
+        if (deviceSession == null) {
+            sendResponse(channel, HttpResponseStatus.NOT_FOUND);
+            return null;
+        }
+
+        Position position = new Position(getProtocolName());
+        position.setDeviceId(deviceSession.getDeviceId());
+
+        JsonObject message = root.getJsonObject("uplink_message");
+        if (message == null) {
+            sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+            return null;
+        }
+
+        JsonObject payload = message.getJsonObject("decoded_payload");
+        // Join messages don't have a payload, but we still need to respond with success to prevent being disabled
+        if (payload == null) {
+            sendResponse(channel, HttpResponseStatus.NO_CONTENT);
+            return null;
+        }
+
+        Network network = new Network();
+        network.setRadioType("lora");
+        network.setCarrier(message.getJsonObject("network_ids").getString("tenant_id"));
```

Comment:
If we don't have cell or wifi info, we should not set network object at all.

Reply:
Not in this PR, but would I be able to add a `LorawanGateway` class, similar to `CellTower` and `WifiAccessPoint`? As TTN does provide a list of gateways that received the message including the rssi and (if published) their lat/lng in the metadata.

Reply:
Let's discuss it separately.